src/train_cmd.cpp

```diff
@@ -1662,19 +1662,31 @@ static bool TrainApproachingCrossing(TileIndex tile)
 	return HasVehicleOnPos(tile_from, &tile, &TrainApproachingCrossingEnum);
 }
 
+/** Check if the crossing should be closed
+ *  @return train on crossing || train approaching crossing || reserved
+ */
+static inline bool CheckLevelCrossing(TileIndex tile)
+{
+	return HasCrossingReservation(tile) || HasVehicleOnPos(tile, nullptr, &TrainOnTileEnum) || TrainApproachingCrossing(tile);
+}
 
 /**
  * Sets correct crossing state
  * @param tile tile to update
  * @param sound should we play sound?
+ * @param force_state force close the crossing due to an adjacent tile
  * @pre tile is a rail-road crossing
  */
-void UpdateLevelCrossing(TileIndex tile, bool sound)
+static void UpdateLevelCrossingTile(TileIndex tile, bool sound, bool force_state = false)
 {
 	assert(IsLevelCrossingTile(tile));
+	bool new_state;
 
-	/* reserved || train on crossing || train approaching crossing */
-	bool new_state = HasCrossingReservation(tile) || HasVehicleOnPos(tile, nullptr, &TrainOnTileEnum) || TrainApproachingCrossing(tile);
+	if (force_state) {
+		new_state = force_state;
+	} else {
+		new_state = CheckLevelCrossing(tile);
+	}
```

Comment:
`bool new_state = force_state ? force_state : CheckLevelCrossing(tile)`

Also, I think `force_state` is a bad name, as it is not only forcing a state, it is also the value of the state if forced. Which is confusing, as you expect it to be only a toggle. Suggestion: `force_barring`. In that case the code can read:

`bool new_state = force_barring ? true : CheckLevelCrossing(tile)`

Reply:
I'd use:
`bool new_state = force_barring || CheckLevelCrossing(tile)`

Reply:
after a bit of consideration, i believe this second case is totally redundant and can be removed. also, that consideration resulted that the helper functions BarCrossing (unused since this patch) and UnbarCrossing (completely unused) can be removed.